### DISCUSSION TITLE
Repeat weird behavior

### DISCUSSION THREAD

**Discussion by mozesa** [louthy/language-ext#1064] [Q&A] [answered]
**Hello**,

would you be so kind as to help me wrap my head around the following behavior.
```csharp
...
from _5 in repeat(
                        // from _51 in SuccessEff(unit)
                        from data in ReadLine(stream)
                        from _52 in Console<RT>.writeLine(data)
                        select unit)
...
```

```csharp
    private static Eff<RT, string> ReadLine(Stream stream) =>
        (from data in Eff(stream.ReadByte)
            from _ in guard(data != -1, Errors.Cancelled)
            select data).FoldUntil(string.Empty, (s, ch) => s + Convert.ToChar(ch), ch => ch == '\0');
```

Output with line `from _51 in SuccessEff(unit)` commented out
------
```text
testData
testData testData
testData testData testData
testData testData testData testData
```

Output with line `from _51 in SuccessEff(unit)`
------
```text
testData
testData 
testData 
testData 
```

I am sending `testData<NUL>` in both cases.

I don't know how it is possible to get stacked data when `ReadLine` being the first linq expression in an `Eff` inside `repeat`.

Thanks for any help.

**Comment by louthy** [comment]
What version are you using?  This looks like an old issue fixed with the latest release

**Reply by mozesa** [reply]
![image](https://user-images.githubusercontent.com/49973426/175290240-c4e2dff2-559d-4c8a-95d7-9aceb349550b.png)

**Comment by mozesa** [comment]
Here is a better example. I hope it will let you find the problem.

```csharp
using System.Text;
using LanguageExt;
using LanguageExt.Common;
using LanguageExt.Effects.Traits;
using LanguageExt.Sys;
using LanguageExt.Sys.Traits;
using static LanguageExt.Prelude;

namespace ConsoleApp;

public class QueueExample<RT>
    where RT : struct,
    HasCancel<RT>,
    HasConsole<RT>,
    HasDirectory<RT>,
    HasFile<RT>
{
    public static Aff<RT, Unit> Main()
    {
        var content = Encoding.ASCII.GetBytes("test\0test\0test\0");
        var memStream = new MemoryStream(100);
        memStream.Write(content, 0, content.Length);
        memStream.Seek(0, SeekOrigin.Begin);

        return repeat(
                   // from _51 in SuccessEff(unit)
                   from ln in (
                       from data in Eff(memStream.ReadByte)
                       from _ in guard(data != -1, Errors.Cancelled)
                       select data).FoldUntil(string.Empty, (s, ch) => s + (char)ch, ch => ch == '\0')
                   from _52 in Console<RT>.writeLine(ln)
                   select unit)
               | @catch(exception => Console<RT>.writeLine(exception.Message));
    }
}
```

Result when this line `// from _51 in SuccessEff(unit)` left commented out.

```text
test
test test
test test test
cancelled
```

Result when this line ` from _51 in SuccessEff(unit)` is enabled.

```text
test
test
test
cancelled
```

**Comment by louthy** [comment, marked as answer]
It’s capturing free variables, which I assume is the issue.  Won’t be able to look into it until next week though.  You may want to look at the EffectsExamples in the Samples folder, it shows how you can use the Pipes feature of Lang-ext for streaming

**Reply by mozesa** [reply]
Thank you for your help. I created an issue from this discussion.

**Reply by mozesa** [reply]
```csharp
static Effect<RT, Unit> mainEffect =>
    repeat(Console<RT>.readKeys)
        | exitIfEscape
        | keyChar
        | words
        | filterEmpty
        | writeLine;
```

I am using the `repeat` for the same purpose as it is in the example.
In the full application I put the content read from stream into a queue, which is processed by a forked task using Pipes.

**Reply by bmazzarol** [reply]
@mozesa and @louthy I have put a fix up for this #1067 have a look. Should resolve this issue.

**Reply by mozesa** [reply]
@bmazzarol I sincerely appreciate your help. Thank you.
